File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
       int n = heights.size();

        int p1 = 0;
        int p2 = n - 1;

        int ans = 0;
        
        while(p1 < p2) {
            int left = heights[p1];
            int right  = heights[p2];

            int dist = p2 - p1;

            ans = max(ans, dist * min(right, left));
            if(left > right) {
                p2--;
            } else if(left < right) {
                p1++;
            } else {
                p1++;
                p2--;
            }
        }

        return ans; 
    }
};

File: Data Structures & Algorithms/daily-temperatures/submission-1.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;

        int[] ans = new int[n];

        int[] st = new int[n];
        int top = -1;

        for(int i = 0; i < n; i++) {
            while(top >= 0 && temperatures[i] > temperatures[st[top]]) {
                ans[st[top]] = i - st[top--];
            }

            st[++top] = i;
        }
        return ans;
    }
}